Limb grower

hamzah-hayat opened the pull request:
:cl: Limb Grower
wip: Added a limb grower machine to the game, uses synthflesh to produce limbs for humans/lizards, currently has sprites and basic code working, but due to the way limbs are initialised all limbs produced are currently invisible (though they still work and can be attached correctly)
/:cl:

I'm making this pull request because I'm having trouble finishing this due to the way body parts work. I've finished spriting and nearly all other code however.
Heres a look at the sprite:
![limb-replicator](https://cloud.githubusercontent.com/assets/11432022/18146899/1a92852e-6fca-11e6-8127-b4f81394c72b.png)

Its also animated, and has animations for filling up when creating limbs
TO DO
- make menus a little better (menu to purge/check chemicals)
- fix limbs appearing invisible
- adding more limbs/uses to the machine (perhaps even making the machine emagable)

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into tgstation/master

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into tgstation/master

hamzah-hayat committed:
Adds basic limb grower to game

hamzah-hayat committed:
Added includes

hamzah-hayat committed:
Can now add synthflesh to grower

Limbgrower can take in synthflesh, starts with 100
uses glass beakers to construct, also changed tech for testing
removed a lot of unneeded code left over from autolathe, including
references to materials

hamzah-hayat committed:
Bugfixes and added to boxstation

fixes bug where limb grower starts with empty reagent container (with
nothing inside it)
also trying to fix limb creating bug

Also added limbgrower to boxstation for easier testing

hamzah-hayat committed:
changed where limb spawns and costs

hamzah-hayat committed:
Fixed menus, added sprites

Also cleaned up all the unneeded code

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge master into Limbgrower

hamzah-hayat committed:
Merge

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

chickenboy10 commented:
Oh bugger, i've messed up really badly

optimumtact commented:
what are you concerned about?

chickenboy10 commented:
It says -126,238 lines in the corner, I guess i haven't merged stuff properly?

optimumtact commented:
Did you touch the map?

if so you have to use the map merger https://tgstation13.org/wiki/Map_Merger as our maps are in a nonstandard format that reduces line conflicts

hamzah-hayat committed:
Fixed the map

AnturK commented:
Good idea.

PKPenguin321 commented:
sounds pretty neat but i wouldnt be shocked if it was denied for the same reason https://github.com/tgstation/tgstation/pull/20055 was denied

maybe make it really inefficient but allow science to upgrade it or something
also post pics of it on the map

AnturK commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,232 @@
+
+/obj/machinery/limbgrower/New()
+	..()
+	create_reagents(0)
```
You sure this is what you want ?

chickenboy10 replied:
I think so?
I'm not quite sure how the reagents system works, so i put that in there (as the same is done on the circuit imprinter)

AnturK commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,232 @@
+							/obj/item/weapon/reagent_containers/glass/beaker = 2,
+							/obj/item/weapon/stock_parts/console_screen = 1)
+
+/obj/machinery/limbgrower/Destroy()
```
No need for this if you're not changing it.

chickenboy10 replied:
True

Shapsy commented:
As long as it takes a bit of setup and it would be more expedient to spend a few seconds looking for your arm that blew off before going to surgery, I don't think this would cause a large degree of power creep, this is a super cool idea

InsaneHyena commented:
Reee, stop taking away robotics job

hamzah-hayat committed:
Probally the cause of the build failure

chickenboy10 commented:
I made this as i've watched many rounds where the roboticist was overwhelmed with a huge number of people coming in with missing limbs due to another double esword meme machine going around and chopping as many limbs of they could.
I want to keep this balanced by requiring synthflesh (a mid level chemical) from chemistry to produce.
Hopefully this is a long term solution to this, as the pull #20055 was literally just adding free limbs to medbay, while this is actually using chemistry to produce the limbs

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master'

hamzah-hayat committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into Limb-Grower

hamzah-hayat committed:
trying to fix the build

InsaneHyena commented:
There's already a limb producer in medbay, tbqh family, it's called genetics. You can even get a limb of your exact skin color, if you like. The only thing that genetics needs are monkey cubes.

Kromgar commented:
You can regrow your limbs by turning into a monkey and back... that really should be fixed

Bawhoppen commented:
don't make it upgradeable, but it needs to require more than synthflesh, make it require synflesh, blood, water, and if we had calcium as a chem I'd say that too but we don't

I'd also say slap one on the white ship aswell as thats supposed to be kind of a medical ship and could be interesting dynamic of fucking off into space to heal their arms after they got shot off by security

Iamgoofball commented:
@Kromgar no way to fix it without it being completely shitcode to the max

RemieRichards commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,232 @@
+								L = new B()
+								world << "Trying to update lizard limb"
+								L.name = "Synthentic Lizard Limb"
+								L.desc = "A synthentic lizard limb that will morph on its first use in surgery"
```
No it won't, you can't use invalid bodyparts in invalid locations
arm in leg slot = failure.

FantasticFwoosh commented:
Literally a few steps away from baymed bio-mass cloning/artificial organ creation.

hamzah-hayat committed:
Limbs are now produced properly

Limbs made properly depending on specias
added plasman as a limb option
addedd a check for beakers putting synthflesh into the grower

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,242 @@
+	//i need to create a body part manually using a set specias dna
+	var/obj/item/bodypart/L
+	var/selected_part
+	world << "category start"
```
debug code left in by mistake?

optimumtact commented on code/game/machinery/limbgrower.dm:
```diff
@@ -0,0 +1,242 @@
+	for(var/obj/item/weapon/reagent_containers/glass/G in component_parts)
+		reagents.maximum_volume += G.volume
+		G.reagents.trans_to(src, G.reagents.total_volume)
+	var/T=1.2
```
can you add a comment to indicate why the value is 1.2 ?

optimumtact commented:
Stale, ask us to reopen this if you intend to continue working on it, or make a new pr from the same branch at this link https://github.com/tgstation/tgstation/compare/master...chickenboy10:Limb-Grower

chickenboy10 commented:
Thats fair enough, ive been working on it between other stuff, but it keeps having problems here and there.
Im also not fully set on what i want to do with it gameplay/balance wise.
ill reopen when im finished.
